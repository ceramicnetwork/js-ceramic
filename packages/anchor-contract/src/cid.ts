import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

const SHA256_CODE = 0x12
const DAG_CBOR_CODE = 0x71
const V1_PROOF_TYPE = 'f(bytes32)'

export function createCidFromHexValue(value: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(value.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}
