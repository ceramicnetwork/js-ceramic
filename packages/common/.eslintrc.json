{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": ["@typescript-eslint", "import"],
  "rules": {
    "@typescript-eslint/no-explicit-any": "off",
    "@typescript-eslint/ban-ts-ignore": "off",
    "@typescript-eslint/ban-ts-comment": "off",
    "import/no-unresolved": "error",
    "import/default": "off"
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [".ts", ".tsx"]
    },
    "import/resolver": {
      "typescript": {
        "extensions": []
      },
      "node": true
    }
  },
  "extends": [
    "eslint:recommended",
    "plugin:@typescript-eslint/eslint-recommended",
    "plugin:@typescript-eslint/recommended",
    "plugin:import/typescript"
  ]
}
