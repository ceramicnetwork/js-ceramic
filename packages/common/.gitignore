node_modules/
coverage/
lib/
src/generated/
