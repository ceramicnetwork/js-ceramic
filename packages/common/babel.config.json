{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "node": "16"
        },
        "useBuiltIns": false,
        "debug": false
      }
    ],
    "@babel/preset-typescript"
  ],
  "plugins": [
    [
      "@babel/plugin-transform-modules-commonjs",
      {
        "allowTopLevelThis": true
      }
    ],
    ["@babel/plugin-proposal-decorators", { "legacy": true }]
  ]
}
