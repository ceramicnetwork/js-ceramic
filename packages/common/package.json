{
  "name": "@ceramicnetwork/common",
  "version": "0.13.0",
  "description": "Ceramic common types and utilities",
  "keywords": [
    "ceramic",
    "types",
    "utilities"
  ],
  "author": "Janko Simonovic <simonovic86@gmail.com>",
  "homepage": "",
  "license": "(Apache-2.0 OR MIT)",
  "directories": {
    "lib": "lib"
  },
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib"
  ],
  "scripts": {
    "test": "./node_modules/.bin/jest --coverage --passWithNoTests",
    "build": "../../node_modules/.bin/tsc -p tsconfig.json",
    "prepublishOnly": "npm run build",
    "prebuild": "npm run clean",
    "lint": "./node_modules/.bin/eslint ./src --ext .js,.jsx,.ts,.tsx",
    "clean": "rm -rf ./lib"
  },
  "dependencies": {
    "@ceramicnetwork/docid": "^0.3.0",
    "cids": "^1.0.2",
    "dag-jose": "^0.3.0",
    "did-resolver": "^2.1.1",
    "lodash.clonedeep": "^4.5.0",
    "loglevel": "^1.7.0",
    "loglevel-plugin-prefix": "^0.8.4",
    "uint8arrays": "^1.1.0"
  },
  "devDependencies": {
    "@babel/core": "^7.9.0",
    "@babel/plugin-proposal-decorators": "^7.10.1",
    "@babel/preset-env": "^7.8.4",
    "@babel/preset-typescript": "^7.9.0",
    "@types/encoding-down": "^5.0.0",
    "@types/events": "^3.0.0",
    "@types/json-schema": "^7.0.5",
    "@types/lodash.clonedeep": "^4.5.6",
    "@types/node": "^13.13.15",
    "@types/prettier": "^2.0.2",
    "@typescript-eslint/eslint-plugin": "^4.6.0",
    "@typescript-eslint/parser": "^4.6.0",
    "babel-jest": "^25.1.0",
    "dids": "^1.0.0",
    "eslint": "^6.8.0",
    "eslint-plugin-jest": "^23.8.2",
    "jest": "^25.1.0",
    "js-sha256": "^0.9.0",
    "json-schema-to-typescript": "^9.1.1",
    "tmp-promise": "^2.0.2",
    "ts-node": "^9.0.0",
    "tsc-watch": "^4.2.9",
    "ipfs-core": "^0.3.0",
    "typescript-json-schema": "^0.42.0"
  },
  "jest": {
    "transformIgnorePatterns": [
      "/node_modules(?!/did-jwt)/"
    ]
  },
  "gitHead": "c58eb9659f5fdd976aca2db6465619c20ba30ff9"
}
