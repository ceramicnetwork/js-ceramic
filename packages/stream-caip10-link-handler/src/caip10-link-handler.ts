import { validateLink } from '@ceramicnetwork/blockchain-utils-validation'
import { Caip10Link } from '@ceramicnetwork/stream-caip10-link'
import {
  AnchorStatus,
  CommitData,
  CommitType,
  Context,
  SignatureStatus,
  StreamConstructor,
  StreamHandler,
  StreamState,
  StreamUtils,
  toLegacyAccountId,
} from '@ceramicnetwork/common'
import { applyAnchorCommit } from '@ceramicnetwork/stream-handler-common'

export class Caip10LinkHandler implements StreamHandler<Caip10Link> {
  get type(): number {
    return Caip10Link.STREAM_TYPE_ID
  }

  get name(): string {
    return Caip10Link.STREAM_TYPE_NAME
  }

  get stream_constructor(): StreamConstructor<Caip10Link> {
    return Caip10Link
  }

  /**
   * Applies commit (genesis|signed|anchor)
   * @param commitData - Commit (with JWS envelope or anchor proof, if available and extracted before application)
   * @param context - Ceramic context
   * @param state - Stream state
   */
  async applyCommit(
    commitData: CommitData,
    context: Context,
    state?: StreamState
  ): Promise<StreamState> {
    if (state == null) {
      return this._applyGenesis(commitData)
    }

    if (StreamUtils.isAnchorCommitData(commitData)) {
      return this._applyAnchor(commitData, state)
    }

    return this._applySigned(commitData, state)
  }

  /**
   * Applies genesis commit
   * @param commitData - Genesis commit
   * @private
   */
  async _applyGenesis(commitData: CommitData): Promise<StreamState> {
    const commit = commitData.commit
    if (commit.data) {
      throw new Error('Caip10Link genesis commit cannot have data')
    }

    const metadata = commit.header
    if (!(metadata.controllers && metadata.controllers.length === 1)) {
      throw new Error('Exactly one controller must be specified')
    }

    // TODO - verify genesis commit
    const state: StreamState = {
      type: Caip10Link.STREAM_TYPE_ID,
      content: null,
      next: {
        content: null,
      },
      metadata,
      signature: SignatureStatus.GENESIS,
      anchorStatus: AnchorStatus.NOT_REQUESTED,
      log: [{ cid: commitData.cid, type: CommitType.GENESIS }],
    }

    return state
  }

  /**
   * Applies signed commit
   * @param commitData - Signed commit
   * @param state - Stream state
   * @private
   */
  async _applySigned(commitData: CommitData, state: StreamState): Promise<StreamState> {
    const commit = commitData.commit
    StreamUtils.assertCommitLinksToState(state, commit)

    let validProof = null
    try {
      validProof = await validateLink(commit.data)
    } catch (e) {
      throw new Error(
        'Error while validating link proof for caip10-link signed commit: ' + e.toString()
      )
    }
    if (!validProof) {
      throw new Error('Invalid proof for signed commit')
    }

    if (
      state.signature !== SignatureStatus.GENESIS &&
      ((state.anchorStatus === AnchorStatus.ANCHORED &&
        validProof.timestamp < state.anchorProof.blockTimestamp) ||
        (state.anchorStatus !== AnchorStatus.ANCHORED &&
          validProof.timestamp < state.next.metadata.lastUpdate))
    ) {
      throw new Error('Invalid commit, proof timestamp too old')
    }

    // TODO: handle CAIP-10 addresses in proof generation of 3id-blockchain-utils
    const account: string = validProof.account || validProof.address

    const legacyAccountCaip10 = toLegacyAccountId(account)

    // We can assume that controllers will always follow the legacy CAIP-10 Link format
    const legacyControllerCaip10 = state.metadata.controllers[0]
    if (!legacyControllerCaip10.includes('@')) {
      throw new Error('Controller is not following the legacy CAIP10 format. Unexpected error.')
    }

    if (legacyAccountCaip10.toLowerCase() !== legacyControllerCaip10.toLowerCase()) {
      throw new Error(
        `Address '${legacyAccountCaip10.toLowerCase()}' used to sign update to Caip10Link doesn't match stream controller '${legacyControllerCaip10.toLowerCase()}'`
      )
    }
    state.log.push({ cid: commitData.cid, type: CommitType.SIGNED })
    return {
      ...state,
      signature: SignatureStatus.SIGNED,
      anchorStatus: AnchorStatus.NOT_REQUESTED,
      next: {
        content: validProof.did,
        metadata: {
          ...state.metadata,
          lastUpdate: validProof.timestamp, // in case there are two updates after each other
        },
      },
    }
  }

  /**
   * Applies anchor commit
   * @param commitData - Anchor commit
   * @param state - Stream state
   * @private
   */
  async _applyAnchor(commitData: CommitData, state: StreamState): Promise<StreamState> {
    return applyAnchorCommit(commitData, state)
  }
}
