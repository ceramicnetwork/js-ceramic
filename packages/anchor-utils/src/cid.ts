import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

export const SHA256_CODE = 0x12
export const KECCAK_256_CODE = 0x1b
export const DAG_CBOR_CODE = 0x71
export const ETH_TX_CODE = 0x93

export function createCidFromHexValue(hexValue: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor(address,bytes32)'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

const V1_PROOF_TYPE = 'f(bytes32)'

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}

/**
 * Converts ETH address to CID
 * @param hash - ETH hash
 */
export function convertEthHashToCid(hexValue: string): CID {
  const multihash = createMultihash(KECCAK_256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, ETH_TX_CODE, multihash)
}
