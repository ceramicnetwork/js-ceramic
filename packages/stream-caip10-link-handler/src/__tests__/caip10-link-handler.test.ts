import { jest } from '@jest/globals'
import { Caip10LinkHandler } from '../caip10-link-handler.js'
import cloneDeep from 'lodash.clonedeep'
import { CID } from 'multiformats/cid'
import { decode as decodeMultiHash } from 'multiformats/hashes/digest'
import { Caip10Link } from '@ceramicnetwork/stream-caip10-link'
import { CeramicApi, CommitType, Context, TestUtils } from '@ceramicnetwork/common'
import sha256 from '@stablelib/sha256'
import * as uint8arrays from 'uint8arrays'
import { AccountId } from 'caip'

const digest = (input: string) =>
  uint8arrays.toString(sha256.hash(uint8arrays.fromString(input)), 'base16')
const hash = (data: string): CID =>
  CID.create(1, 0x12, decodeMultiHash(Buffer.from('1220' + digest(data), 'hex')))

const FAKE_CID_1 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_5 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlse')
const FAKE_CID_6 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjleu')
const FAKE_CID_7 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlwu')
const FAKE_CID_8 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts66jqcuam6bmye2pb54adnrtccjlwu')

const ACCOUNT = '0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1'

const COMMITS = {
  genesis: { header: { controllers: [ACCOUNT.toLowerCase()], family: 'caip10-eip155:1' } },
  r1: {
    desiredContent: {
      version: 2,
      type: 'ethereum-eoa',
      message:
        'Create a new account link to your identity.\n' +
        '\n' +
        'did:3:testdid1 \n' +
        'Timestamp: 1608116721',
      signature:
        '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e83078c907ccfaf9bd0de82bd5e57fe7884d9cb8a9303f5d92ea4df0d1c',
      account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
      timestamp: 1608116721,
    },
    commit: {
      data: {
        version: 2,
        type: 'ethereum-eoa',
        message:
          'Create a new account link to your identity.\n' +
          '\n' +
          'did:3:testdid1 \n' +
          'Timestamp: 1608116721',
        signature:
          '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e83078c907ccfaf9bd0de82bd5e57fe7884d9cb8a9303f5d92ea4df0d1c',
        account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
        timestamp: 1608116721,
      },
      id: FAKE_CID_1,
      prev: FAKE_CID_1,
    },
  },
  r2: { commit: { proof: FAKE_CID_4, id: FAKE_CID_1, prev: FAKE_CID_2 } },
  proof: {
    value: {
      blockNumber: 123456,
      blockTimestamp: 1666728832,
      chainId: 'fakechain:123',
    },
  },
}

describe('Caip10LinkHandler', () => {
  let context: Context
  let handler: Caip10LinkHandler

  beforeEach(() => {
    handler = new Caip10LinkHandler()
    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = rec
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = clone
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        },
      },
    }

    const api = {
      getSupportedChains: jest.fn(async () => {
        return ['fakechain:123']
      }),
    }
    context = {
      ipfs: ipfs,
      anchorService: null,
      api: api as unknown as CeramicApi,
    }
  })

  it('is constructed correctly', async () => {
    expect(handler.name).toEqual('caip10-link')
  })

  it('makes genesis commit correctly', async () => {
    const commit = Caip10Link.makeGenesis(
      new AccountId({ address: '0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8', chainId: 'eip155:1' })
    )
    expect(commit).toEqual(COMMITS.genesis)
  })

  it('throws an error if genesis commit has data', async () => {
    const commitWithData = { ...COMMITS.genesis, data: {} }
    const genesisWithData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: commitWithData }
    await expect(handler.applyCommit(genesisWithData, context)).rejects.toThrow(/cannot have data/)
  })

  it('throws an error if genesis commit has no controllers specified', async () => {
    const commitWithoutControllers = cloneDeep(COMMITS.genesis)
    delete commitWithoutControllers.header.controllers
    const genesisWithoutControllers = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: commitWithoutControllers,
    }
    await expect(handler.applyCommit(genesisWithoutControllers, context)).rejects.toThrow(
      /Exactly one controller must be specified/i
    )
  })

  it('throws an error if genesis commit has more than one controller', async () => {
    const commitWithMultipleControllers = cloneDeep(COMMITS.genesis)
    commitWithMultipleControllers.header.controllers.push(
      '0x25954ef14cebbc9af3d79876489a9cfe87043f20@eip155:1'
    )
    const genesisWithMultipleControllers = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: commitWithMultipleControllers,
    }
    await expect(handler.applyCommit(genesisWithMultipleControllers, context)).rejects.toThrow(
      /Exactly one controller must be specified/i
    )
  })

  it('applies genesis commit correctly', async () => {
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    const state = await handler.applyCommit(genesisCommitData, context)
    expect(state).toMatchSnapshot()
  })

  it('makes update commit correctly', async () => {
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    const state = await handler.applyCommit(genesisCommitData, context)
    const state$ = TestUtils.runningState(state)
    const stream = new Caip10Link(state$, context)
    const commit = await stream.makeCommit(COMMITS.r1.desiredContent)
    // Have to compare the 'id' and 'prev' CIDs manually (with toString()) otherwise jest gets
    // confused by Symbol(@ipld/js-cid/CID)
    expect(commit.data).toEqual(COMMITS.r1.commit.data)
    expect(commit.id.toString()).toEqual(COMMITS.r1.commit.id.toString())
    expect(commit.prev.toString()).toEqual(COMMITS.r1.commit.prev.toString())
  })

  it('applies signed commit correctly', async () => {
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    let state = await handler.applyCommit(genesisCommitData, context)
    const signedCommitData = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit: COMMITS.r1.commit }
    state = await handler.applyCommit(signedCommitData, context, state)
    expect(state).toMatchSnapshot()
  })

  it('throws an error of the proof is invalid', async () => {
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    const state = await handler.applyCommit(genesisCommitData, context)
    const badRecord = cloneDeep(COMMITS.r1.commit)
    badRecord.data.signature =
      '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e8302348762387462387648726346877884d9cb8a9303f5d92ea4df0d1c'
    const badCommitData = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit: badRecord }
    await expect(handler.applyCommit(badCommitData, context, state)).rejects.toThrow(
      /Invalid proof/i
    )
  })

  it("throws an error of the proof doesn't match the controller", async () => {
    const badAddressGenesis = cloneDeep(COMMITS.genesis)
    badAddressGenesis.header.controllers = ['0xffffffffffffffffffffffffffffffffffffffff@eip155:1']
    const badAddressGenesisData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: badAddressGenesis,
    }
    const state = await handler.applyCommit(badAddressGenesisData, context)
    const signedCommitData = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit: COMMITS.r1.commit }
    await expect(handler.applyCommit(signedCommitData, context, state)).rejects.toThrow(
      `Address '${COMMITS.r1.commit.data.account.toLowerCase()}' used to sign update to Caip10Link doesn't match stream controller '${badAddressGenesis.header.controllers[0].toLowerCase()}'`
    )
  })

  it('fails to apply commit with invalid prev link', async () => {
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    const state = await handler.applyCommit(genesisCommitData, context)
    const commit = { ...COMMITS.r1.commit }
    commit.prev = FAKE_CID_3
    const signedCommitData = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit }
    await expect(handler.applyCommit(signedCommitData, context, state)).rejects.toThrow(
      /Commit doesn't properly point to previous commit in log/
    )
  })

  it('fails to apply commit with invalid id property', async () => {
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    const state = await handler.applyCommit(genesisCommitData, context)
    const commit = { ...COMMITS.r1.commit }
    commit.id = FAKE_CID_3
    const signedCommitData = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit }
    await expect(handler.applyCommit(signedCommitData, context, state)).rejects.toThrow(
      /Invalid genesis CID in commit/
    )
  })

  it('applies anchor commit correctly', async () => {
    // create signed commit
    await context.ipfs.dag.put(COMMITS.r1.commit, FAKE_CID_2)
    // create anchor commit
    await context.ipfs.dag.put(COMMITS.r2.commit, FAKE_CID_3)
    // create anchor proof
    await context.ipfs.dag.put(COMMITS.proof, FAKE_CID_4)

    // Apply genesis
    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: COMMITS.genesis }
    let state = await handler.applyCommit(genesisCommitData, context)
    // Apply signed record
    const signedCommitData = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit: COMMITS.r1.commit }
    state = await handler.applyCommit(signedCommitData, context, state)
    // Apply anchor record
    const anchorCommitData = {
      cid: FAKE_CID_3,
      type: CommitType.ANCHOR,
      commit: COMMITS.r2.commit,
      proof: COMMITS.proof.value,
    }
    state = await handler.applyCommit(anchorCommitData, context, state)
    expect(state).toMatchSnapshot()
  })

  it('Should not allow replay attack', async () => {
    const commits = {
      genesis: {
        header: {
          controllers: ['0x0544DcF4fcE959C6C4F3b7530190cB5E1BD67Cb8@eip155:1'],
          family: 'caip10-eip155:1',
        },
      },
      r1: {
        data: {
          version: 2,
          type: 'ethereum-eoa',
          message:
            'Create a new account link to your identity.\n' +
            '\n' +
            'did:3:testdid1 \n' +
            'Timestamp: 1608116721',
          signature:
            '0xc6a5f50945bc7b06320b66cfe144e2b571391c88827eed0490f7f8e5e8af769c4246e27e83078c907ccfaf9bd0de82bd5e57fe7884d9cb8a9303f5d92ea4df0d1c',
          account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
          timestamp: 1608116721,
        },
        id: FAKE_CID_1,
        prev: FAKE_CID_1,
      },
      r2: { proof: FAKE_CID_4, id: FAKE_CID_1, prev: FAKE_CID_2 },
      r2proof: {
        value: {
          blockNumber: 123456,
          blockTimestamp: 1608116723,
          chainId: 'fakechain:123',
        },
      },
      r3: {
        data: {
          version: 2,
          type: 'ethereum-eoa',
          message:
            'Create a new account link to your identity.\n' +
            '\n' +
            'did:3:testdid2 \n' +
            'Timestamp: 1608116725',
          signature:
            '0xe1bcc3f3b67ae303cb95db9d2d266f95984fd76bf3a4452dbb64ad4d3941998f4cb37f85197c74f9fb5cdf33e9042949a5452a204db2d48d85929406f64aac811b',
          account: '0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1',
          timestamp: 1608116725,
        },
        id: FAKE_CID_1,
        prev: FAKE_CID_3,
      },
      r4: { proof: FAKE_CID_7, id: FAKE_CID_1, prev: FAKE_CID_5 },
      r4proof: {
        value: {
          blockNumber: 123456,
          blockTimestamp: 1608116727,
          chainId: 'fakechain:123',
        },
      },
    }
    await context.ipfs.dag.put(commits.r2proof, FAKE_CID_4)
    await context.ipfs.dag.put(commits.r4proof, FAKE_CID_7)

    const genesisCommitData = { cid: FAKE_CID_1, type: CommitType.GENESIS, commit: commits.genesis }
    let state = await handler.applyCommit(genesisCommitData, context)
    const signedCommitData_1 = { cid: FAKE_CID_2, type: CommitType.SIGNED, commit: commits.r1 }
    state = await handler.applyCommit(signedCommitData_1, context, state)
    const anchorCommitData_1 = {
      cid: FAKE_CID_3,
      type: CommitType.ANCHOR,
      commit: commits.r2,
      proof: commits.r2proof.value,
      timestamp: commits.r2proof.value.blockTimestamp,
    }
    state = await handler.applyCommit(anchorCommitData_1, context, state)
    expect(state.content).toEqual('did:3:testdid1')
    const signedCommitData_2 = { cid: FAKE_CID_5, type: CommitType.SIGNED, commit: commits.r3 }
    state = await handler.applyCommit(signedCommitData_2, context, state)

    // create a fake update based on the r1 data to try a replay attack
    const r4 = {
      data: commits.r1.data,
      id: FAKE_CID_1,
      prev: FAKE_CID_5,
    }
    const signedCommitData_3 = { cid: FAKE_CID_8, type: CommitType.SIGNED, commit: r4 }
    await expect(handler.applyCommit(signedCommitData_3, context, state)).rejects.toThrow(
      'Invalid commit, proof timestamp too old'
    )

    const anchorCommitData_2 = {
      cid: FAKE_CID_6,
      type: CommitType.ANCHOR,
      commit: commits.r4,
      proof: commits.r4proof.value,
    }
    state = await handler.applyCommit(anchorCommitData_2, context, state)
    expect(state.content).toEqual('did:3:testdid2')

    // create a fake update based on the r1 data to try a replay attack
    const r5 = {
      data: commits.r1.data,
      id: FAKE_CID_1,
      prev: FAKE_CID_6,
    }
    const signedCommitData_4 = { cid: FAKE_CID_8, type: CommitType.SIGNED, commit: r5 }
    await expect(handler.applyCommit(signedCommitData_4, context, state)).rejects.toThrow(
      'Invalid commit, proof timestamp too old'
    )
  })
})
