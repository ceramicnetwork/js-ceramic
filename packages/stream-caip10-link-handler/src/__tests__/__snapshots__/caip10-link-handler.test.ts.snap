// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caip10LinkHandler applies anchor commit correctly 1`] = `
{
  "anchorProof": {
    "chainId": "fakechain:123",
  },
  "anchorStatus": 3,
  "content": "did:3:testdid1",
  "log": [
    {
      "cid": {
        "/": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
      },
      "timestamp": 1666728832,
      "type": 0,
    },
    {
      "cid": {
        "/": "bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu",
      },
      "timestamp": 1666728832,
      "type": 1,
    },
    {
      "cid": {
        "/": "bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu",
      },
      "timestamp": 1666728832,
      "type": 2,
    },
  ],
  "metadata": {
    "controllers": [
      "0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1",
    ],
    "family": "caip10-eip155:1",
    "lastUpdate": 1608116721,
  },
  "signature": 2,
  "type": 1,
}
`;

exports[`Caip10LinkHandler applies genesis commit correctly 1`] = `
{
  "anchorStatus": 0,
  "content": null,
  "log": [
    {
      "cid": {
        "/": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
      },
      "type": 0,
    },
  ],
  "metadata": {
    "controllers": [
      "0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1",
    ],
    "family": "caip10-eip155:1",
  },
  "next": {
    "content": null,
  },
  "signature": 0,
  "type": 1,
}
`;

exports[`Caip10LinkHandler applies signed commit correctly 1`] = `
{
  "anchorStatus": 0,
  "content": null,
  "log": [
    {
      "cid": {
        "/": "bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu",
      },
      "type": 0,
    },
    {
      "cid": {
        "/": "bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu",
      },
      "type": 1,
    },
  ],
  "metadata": {
    "controllers": [
      "0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1",
    ],
    "family": "caip10-eip155:1",
  },
  "next": {
    "content": "did:3:testdid1",
    "metadata": {
      "controllers": [
        "0x0544dcf4fce959c6c4f3b7530190cb5e1bd67cb8@eip155:1",
      ],
      "family": "caip10-eip155:1",
      "lastUpdate": 1608116721,
    },
  },
  "signature": 2,
  "type": 1,
}
`;
